File: pipeline-visualizer/src/utils/dataGenerator.js
const BUSINESS_NAMES = [
  'Tech Solutions Inc',
  'Global Logistics Ltd',
  'Green Energy Co',
  'Precision Manufacturing',
  'Digital Marketing Agency',
  'Wholesale Distributors Inc'
];

const CURRENCIES = ['USD', 'SGD', 'EUR', 'GBP', 'AED'];

const TRANSACTION_CATEGORIES = {
  credit: ['sales_revenue', 'customer_payment', 'service_revenue', 'contract_payment', 'subscription_revenue'],
  debit: ['payroll', 'vendor_payment', 'rent_lease', 'software_subscription', 'marketing']
};

const PROFILE_CONFIGS = {
  balanced: {
    creditBias: 0.62,
    amountMultiplier: 1,
    spanDays: 45,
    lookbackDays: 60,
    crossBorderRate: 0.25,
    descriptionPrefix: 'Growth and expansion'
  },
  negative: {
    creditBias: 0.3,
    amountMultiplier: 1.25,
    spanDays: 60,
    lookbackDays: 75,
    crossBorderRate: 0.5,
    descriptionPrefix: 'Liquidity pressure'
  },
  neutral: {
    creditBias: 0.5,
    amountMultiplier: 0.85,
    spanDays: 30,
    lookbackDays: 30,
    crossBorderRate: 0.12,
    descriptionPrefix: 'Steady-state operations'
  }
};

const COUNTERPARTIES = [
  'Microsoft Corporation',
  'Amazon Web Services Inc',
  'Google LLC',
  'Adobe Inc',
  'PayPal Holdings Inc',
  'Stripe Inc'
];

const NOTES = [
  'Quarterly subscription',
  'Customer milestone payment',
  'Vendor settlement',
  'Payroll run',
  'Infrastructure investment'
];

const randomInt = (min, max) => Math.floor(Math.random() * (max - min + 1)) + min;
const randomChoice = (arr) => arr[randomInt(0, arr.length - 1)];

const buildAccountInfo = () => ({
  account_number: `${randomInt(10000000, 99999999)}`,
  routing_number: `${randomInt(100000000, 999999999)}`,
  bank_name: randomChoice(['First National Bank', 'Commerce Bank', 'Enterprise Bank']),
  account_type: randomChoice(['business_checking', 'commercial_account'])
});

const chooseTransactionType = (config) =>
  Math.random() < config.creditBias ? 'credit' : 'debit';

const buildTransaction = (index, transactionType, referenceDate, customerName, customerId, config, accountInfo) => {
  const category = randomChoice(TRANSACTION_CATEGORIES[transactionType]);
  const base = transactionType === 'credit' ? 15000 : 8000;
  const variance = transactionType === 'credit' ? 10000 : 5000;
  const amount = (base + Math.random() * variance) * config.amountMultiplier;
  const signedAmount = transactionType === 'debit' ? -Math.abs(amount) : amount;
  const timestamp = new Date(referenceDate.getTime() + randomInt(0, config.spanDays) * 86400000);
  const description = `${config.descriptionPrefix} - ${NOTES[index % NOTES.length]}`;
  const tags = transactionType === 'debit' ? ['card_spend'] : ['incoming_payment'];
  if (Math.random() < config.crossBorderRate) {
    tags.push('cross_border');
  }

  const beneficiary = randomChoice(COUNTERPARTIES);

  return {
    transaction_date: timestamp.toISOString().split('T')[0],
    amount: Number(signedAmount.toFixed(2)),
    currency: randomChoice(CURRENCIES),
    transaction_type: transactionType,
    description,
    transaction_category: category,
    tags,
    remittance_information: description,
    date: timestamp.toISOString().split('T')[0],
    type: transactionType,
    originator: customerName,
    beneficiary,
    counterparty: beneficiary,
    debtor: customerName,
    creditor: beneficiary,
    debtor_account: accountInfo.account_number,
    creditor_account: accountInfo.routing_number,
    initiating_party: customerId,
    ultimate_debtor: customerName,
    ultimate_creditor: beneficiary,
    merchant: beneficiary,
    originator_account: accountInfo.account_number,
    beneficiary_account: accountInfo.routing_number
  };
};

export const generateDataset = (profile = 'balanced', transactionCount = 40) => {
  const config = PROFILE_CONFIGS[profile] ?? PROFILE_CONFIGS.balanced;
  const safeCount = Math.max(8, Math.min(5000, transactionCount));
  const referenceDate = new Date(Date.now() - config.lookbackDays * 24 * 3600 * 1000);
  const customerName = randomChoice(BUSINESS_NAMES);
  const customerId = `${customerName.split(' ')[0].slice(0, 3).toUpperCase()}_${randomInt(100, 999)}`;
  const accountInfo = buildAccountInfo();

  const transactions = [];
  for (let idx = 0; idx < safeCount; idx += 1) {
    const transactionType = chooseTransactionType(config);
    transactions.push(
      buildTransaction(idx, transactionType, referenceDate, customerName, customerId, config, accountInfo)
    );
  }

  return {
    customer_id: customerId,
    name: customerName,
    email: `contact@${customerName.toLowerCase().replace(/\s+/g, '')}.com`,
    phone: `+1-${randomInt(200, 999)}-${randomInt(1000, 9999)}`,
    currency: randomChoice(CURRENCIES),
    transactions,
    account_info: accountInfo,
    profile,
    generated_at: new Date().toISOString()
  };
};

File: app/data_generator.py
"""
Lightweight dataset factory for the prompt engine UI
"""
from __future__ import annotations

import random
from datetime import datetime, timedelta
from typing import Any, Dict, List


class DynamicDatasetGenerator:
    def __init__(self) -> None:
        self.business_names = [
            'Tech Solutions Inc', 'Global Logistics Ltd', 'Green Energy Co',
            'Precision Manufacturing', 'Digital Marketing Agency', 'Wholesale Distributors Inc'
        ]
        self.currencies = ['USD', 'SGD', 'EUR', 'GBP', 'AED']
        self.transaction_categories = {
            'credit': [
                'sales_revenue', 'customer_payment', 'service_revenue',
                'contract_payment', 'subscription_revenue'
            ],
            'debit': [
                'payroll', 'vendor_payment', 'rent_lease',
                'software_subscription', 'marketing'
            ]
        }
        self.profile_configs = {
            'balanced': {
                'credit_bias': 0.62,
                'amount_multiplier': 1.0,
                'span_days': 45,
                'lookback_days': 60,
                'cross_border_rate': 0.25,
                'description_prefix': 'Growth and expansion'
            },
            'negative': {
                'credit_bias': 0.3,
                'amount_multiplier': 1.25,
                'span_days': 60,
                'lookback_days': 75,
                'cross_border_rate': 0.5,
                'description_prefix': 'Liquidity pressure'
            },
            'neutral': {
                'credit_bias': 0.5,
                'amount_multiplier': 0.85,
                'span_days': 30,
                'lookback_days': 30,
                'cross_border_rate': 0.12,
                'description_prefix': 'Steady-state operations'
            }
        }
        self.counterparties = [
            'Microsoft Corporation', 'Amazon Web Services Inc', 'Google LLC',
            'Adobe Inc', 'PayPal Holdings Inc', 'Stripe Inc'
        ]
        self.notes = [
            'Quarterly subscription', 'Customer milestone payment',
            'Vendor settlement', 'Payroll run', 'Infrastructure investment'
        ]

    def generate(
        self,
        profile: str = 'balanced',
        transaction_count: int = 30
    ) -> Dict[str, Any]:
        config = self.profile_configs.get(profile, self.profile_configs['balanced'])
        transaction_count = max(8, min(200, transaction_count))
        reference_date = datetime.utcnow() - timedelta(days=config['lookback_days'])
        customer_name = random.choice(self.business_names)
        customer_id = f"{customer_name.split()[0][:3].upper()}_{random.randint(100, 999)}"

        account_info = self._build_account_info()
        transactions: List[Dict[str, Any]] = []
        for idx in range(transaction_count):
            txn_type = self._choose_transaction_type(config)
            transaction = self._build_transaction(
                idx,
                txn_type,
                reference_date,
                customer_name,
                customer_id,
                config,
                account_info
            )
            transactions.append(transaction)

        return {
            'customer_id': customer_id,
            'name': customer_name,
            'email': f"contact@{customer_name.lower().replace(' ', '')}.com",
            'phone': f"+1-{random.randint(200, 999)}-{random.randint(1000, 9999)}",
            'currency': random.choice(self.currencies),
            'transactions': transactions,
            'account_info': account_info,
            'profile': profile,
            'generated_at': datetime.utcnow().isoformat()
        }

    def _choose_transaction_type(self, config: Dict[str, Any]) -> str:
        return 'credit' if random.random() < config['credit_bias'] else 'debit'

    def _build_transaction(
        self,
        index: int,
        transaction_type: str,
        base_date: datetime,
        customer_name: str,
        customer_id: str,
        config: Dict[str, Any],
        account_info: Dict[str, str]
    ) -> Dict[str, Any]:
        category = random.choice(self.transaction_categories[transaction_type])
        amount = self._random_amount(config, transaction_type)
        currency = random.choice(self.currencies)
        timestamp = base_date + timedelta(days=random.randint(0, config['span_days']))

        description = f"{config['description_prefix']} - {self.notes[index % len(self.notes)]}"
        tags = ['card_spend'] if transaction_type == 'debit' else ['incoming_payment']
        if random.random() < config['cross_border_rate']:
            tags.append('cross_border')

        beneficiary = random.choice(self.counterparties)

        return {
            'transaction_date': timestamp.date().isoformat(),
            'amount': round(amount, 2),
            'currency': currency,
            'transaction_type': transaction_type,
            'description': description,
            'transaction_category': category,
            'tags': tags,
            'remittance_information': description,
            'date': timestamp.date().isoformat(),
            'type': transaction_type,
            'originator': customer_name,
            'beneficiary': beneficiary,
            'counterparty': beneficiary,
            'debtor': customer_name,
            'creditor': beneficiary,
            'debtor_account': account_info['account_number'],
            'creditor_account': account_info['routing_number'],
            'initiating_party': customer_id,
            'ultimate_debtor': customer_name,
            'ultimate_creditor': beneficiary,
            'merchant': beneficiary,
            'originator_account': account_info['account_number'],
            'beneficiary_account': account_info['routing_number']
        }

    def _random_amount(self, config: Dict[str, Any], transaction_type: str) -> float:
        base = 15000 if transaction_type == 'credit' else 8000
        variance = 10000 if transaction_type == 'credit' else 5000
        amount = (base + random.uniform(0, variance)) * config['amount_multiplier']
        return amount if transaction_type == 'credit' else -abs(amount)

    def _build_account_info(self) -> Dict[str, str]:
        return {
            'account_number': f"{random.randint(10000000, 99999999)}",
            'routing_number': f"{random.randint(100000000, 999999999)}",
            'bank_name': random.choice(['First National Bank', 'Commerce Bank', 'Enterprise Bank']),
            'account_type': random.choice(['business_checking', 'commercial_account'])
        }

